### PR TITLE
Implement proper support for nested complex env values

### PR DESCRIPTION
**Pull request by hramezani**
Port https://github.com/pydantic/pydantic/pull/4216

### CODE REVIEW

File: pydantic_settings/sources.py [hunk 1]
```diff
@@ -288,6 +288,16 @@ def _field_is_complex(self, field: FieldInfo) -> Tuple[bool, bool]:
 
         return True, allow_parse_failure
 
+    @staticmethod
+    def next_field(field: Optional[FieldInfo], key: str) -> Optional[FieldInfo]:
+        if not field or origin_is_union(get_origin(field.annotation)):
```
Comment (samuelcolvin) [inline pydantic_settings/sources.py:293]: can you add a docstring?

File: pydantic_settings/sources.py [hunk 2]
```diff
@@ -307,8 +317,24 @@ def explode_env_vars(
...
+                target_field = self.next_field(target_field, key)
                 env_var = env_var.setdefault(key, {})
+
+            # get proper field with last_key
+            target_field = self.next_field(target_field, last_key)
+
+            # check if env_val maps to a complex field and if so, parse the env_val
+            if target_field and env_val:
+                is_complex, allow_json_failure = self._field_is_complex(target_field)
+                if is_complex:
```
Comment (samuelcolvin) [inline pydantic_settings/sources.py:332]: do we have a way to disable json parsing?

**Review by samuelcolvin — COMMENTED** [review]
otherwise I think looks good, but I'm a bit unclear how this works, will be easier to review when `next_field` has a docstring.

File: pydantic_settings/sources.py (hunk 1 above)
Comment (hramezani) [reply pydantic_settings/sources.py:312]: added!

File: pydantic_settings/sources.py (hunk 2 above)
Comment (hramezani) [reply pydantic_settings/sources.py:351]: I've tried to disable it here but it doesn't work
I haven't looked deeply on it but seems we need it here

**Comment by samuelcolvin** [comment]
thanks.

**Pull request merged**